File: BP.py
import os
from time import sleep

def limpar_tela():
    os.system('cls' if os.name == 'nt' else 'clear')

def obter_float(prompt):
    while True:
        try:
            v = float(input(prompt))
            if v < 0: raise ValueError
            return v
        except ValueError:
            print("Entrada inválida! Digite um número positivo.")

def coletar_dados(titulo, categorias=True):
    itens = []
    limpar_tela()
    print(f"--- {titulo} ---")
    qtd = int(obter_float(f"Quantidade de itens para {titulo}: "))
    
    # Mapeamento para os índices
    print("\nTags especiais: [1]Estoque [2]Contas a Receber [3]Disponibilidade [4]Fornecedores [0]Nenhum")
    
    for i in range(qtd):
        nome = input(f"\nNome do item {i+1}: ")
        valor = obter_float(f"Valor (R$): ")
        tipo = int(obter_float("Tipo (1-Circulante, 2-Não Circulante): ")) if categorias else 0
        tag = int(obter_float("Tag especial (0-4): "))
        itens.append({'nome': nome, 'valor': valor, 'tipo': tipo, 'tag': tag})
    return itens

def executar_dre():
    d = {}
    print("\n" + "="*20)
    print("   PASSO 1: DRE")
    print("="*20)
    
    # Receitas
    d['rob'] = obter_float("Receita Bruta (Vendas + Serviços): ")
    d['desc'] = obter_float("(-) Desconto sobre Vendas/Abatimentos: ")
    d['rol'] = d['rob'] - d['desc']
    print(f"  > Receita Líquida: R$ {d['rol']:.2f}")

    # Custos (Separados conforme solicitado)
    d['cmv'] = obter_float("(-) CMV (Custo de Mercadorias Vendidas): ")
    d['csp'] = obter_float("(-) CSP (Custo de Serviços Prestados): ")
    d['lucro_bruto'] = d['rol'] - (d['cmv'] + d['csp'])
    print(f"  > Lucro Bruto: R$ {d['lucro_bruto']:.2f}")

    # Despesas e Outros Resultados
    d['desp_op'] = obter_float("(-) Despesas Operacionais (Adm/Vendas/Salários): ")
    d['lucro_op'] = d['lucro_bruto'] - d['desp_op']
    print(f"  > Lucro Operacional: R$ {d['lucro_op']:.2f}")

    d['res_fin'] = obter_float("(+) Resultado Financeiro Líquido: ")
    d['outras_rec'] = obter_float("(+) Outras Receitas (Ex: Venda Imobilizado): ")
    
    # Cálculo Final
    d['lucro_liquido'] = d['lucro_op'] + d['res_fin'] + d['outras_rec']
    
    # Dado necessário para o cálculo do PMP (Prazo Médio de Pagamento)
    d['compras'] = obter_float("Total de Compras de Estoque no período: ")
    
    print("\n" + "-"*30)
    print(f"LUCRO LÍQUIDO FINAL: R$ {d['lucro_liquido']:.2f}")
    print("-"*30)
    
    return d

def calcular_indicadores(dre, ativos, passivos, pl_total):
    # Extração de valores pelas Tags
    extrair = lambda lista, tag: sum(item['valor'] for item in lista if item.get('tag') == tag)
    
    estoque = extrair(ativos, 1)
    receber = extrair(ativos, 2)
    disponiv = extrair(ativos, 3)
    forneced = extrair(passivos, 4)
    
    ac = sum(i['valor'] for i in ativos if i['tipo'] == 1)
    pc = sum(i['valor'] for i in passivos if i['tipo'] == 1)
    at = sum(i['valor'] for i in ativos)
    pt = sum(i['valor'] for i in passivos)

    # Dicionário de resultados
    res = {}
    try:
        # Liquidez
        res['ILC'] = ac / pc
        res['ILS'] = (ac - estoque) / pc
        res['ILI'] = disponiv / pc
        # Atividade (Ano 360 dias)
        res['Giro Estoque'] = dre['cmv'] / estoque if estoque > 0 else 0
        res['PMR'] = (receber * 360) / dre['rol']
        res['PMP'] = (forneced * 360) / dre['compras']
        res['Giro Ativo'] = dre['rol'] / at
        # Endividamento e Rentabilidade
        res['Endiv. Geral'] = (pt / at) * 100
        res['M. Bruta'] = (dre['lucro_bruto'] / dre['rol']) * 100
        res['M. Operacional'] = (dre['lucro_op'] / dre['rol']) * 100
        res['M. Líquida'] = (dre['lucro_liquido'] / dre['rol']) * 100
    except ZeroDivisionError:
        pass
    return res

def main():
    dados_dre = executar_dre()
    input("\nDRE Concluída. Pressione Enter para o Balanço...")
    
    ativos = coletar_dados("ATIVOS")
    passivos = coletar_dados("PASSIVOS")
    
    itens_pl = coletar_dados("PATRIMÔNIO LÍQUIDO", categorias=False)
    # --- CORREÇÃO AQUI: Somar os valores das listas ---
    total_ativos = sum(item['valor'] for item in ativos)
    total_passivos = sum(item['valor'] for item in passivos)
    pl_total = sum(item['valor'] for item in itens_pl) + dados_dre['lucro_liquido']
    
    # Agora o cálculo funciona pois estamos usando os totais numéricos
    diferenca = total_ativos - (total_passivos + pl_total)
    print("\n", 45*'=')
    print("BALANÇO PATRIMONIAL")
    print(45*'=')
    print(f"Total Ativos: R$ {total_ativos:.2f}")
    print(f"Total Passivo + PL: R$ {total_passivos + pl_total:.2f}")
    print(f"Diferença Balanço: R$ {diferenca:.2f}")
    if (diferenca > 0): print("Ativos maior que passivos.")
    if (diferenca < 0): print("Passivos maior que ativos.")
    if (diferenca == 0): print("Balanço patrimonial equilibrado.")
    input("\nBP Concluído. Pressione Enter para o Indicadores...")

    ind = calcular_indicadores(dados_dre, ativos, passivos, pl_total)
    
    limpar_tela()
    print("="*45)
    print(f"{'RELATÓRIO DE INDICADORES':^45}")
    print("="*45)
    
    secoes = {
        "LIQUIDEZ": ["ILC", "ILS", "ILI"],
        "ATIVIDADE": ["Giro Estoque", "PMR", "PMP", "Giro Ativo"],
        "RENTABILIDADE": ["M. Bruta", "M. Operacional", "M. Líquida", "Endiv. Geral"]
    }

    for secao, chaves in secoes.items():
        print(f"\n> {secao}")
        for c in chaves:
            sufixo = "%" if "%" in c or "M." in c or "Endiv" in c else ""
            print(f"  {c:<18}: {ind.get(c, 0):>10.2f}{sufixo}")
    
    print("\n" + "="*45)

if __name__ == "__main__":
    main()

File: BP.c
#include <stdio.h>
#include <stdlib.h>
#include <windows.h>
#include <locale.h>

typedef struct{
    char N[36];
    float Valor;
    int T;
}Dados;

int Maior_0(int V);
void Titulo1();
void Titulo2();
void Titulo3();
void Circulante(int X, Dados B[X]);
void Nao_Circulante(int X, Dados B[X]);

// NOVO
void Mostrar_DRE(int R, Dados Receita[R], int D, Dados Despesa[D]);

int main()
{
    setlocale(LC_ALL, "Portuguese");

    int QA, QP, QV, QR, QD;
    int i, j;

    float SA = 0, SP = 0, Soma_PL = 0;

    // =========================
    // BALANÇO PATRIMONIAL
    // =========================

    do
    {
        printf("=======================\n");
        printf("  BALANCO PATRIMONIAL\n");
        printf("=======================\n\n");

        printf("- Quantidade de ativos: ");
        scanf("%d", &QA);

        printf("- Quantidade de passivos: ");
        scanf("%d", &QP);

        printf("- Quantidade de PL: ");
        scanf("%d", &QV);

    }while((QA < 1) || (QP < 1) || (QV < 1));

    Dados A[QA], P[QP], PL[QV];

    system("cls");

    Titulo1();

    for(i = 0; i < QA; i++)
    {
        printf("Nome: ");
        scanf(" %[^\n]s", A[i].N);

        printf("Valor(R$): ");
        scanf("%f", &A[i].Valor);

        printf("Tipo (1 circulante / 2 nao): ");
        scanf("%d", &A[i].T);

        SA += A[i].Valor;
        system("cls");
    }

    Titulo2();

    for(i = 0; i < QP; i++)
    {
        printf("Nome: ");
        scanf(" %[^\n]s", P[i].N);

        printf("Valor(R$): ");
        scanf("%f", &P[i].Valor);

        printf("Tipo (1 circulante / 2 nao): ");
        scanf("%d", &P[i].T);

        SP += P[i].Valor;
    }

    Titulo3();

    for(i = 0; i < QV; i++)
    {
        printf("Nome: ");
        scanf(" %[^\n]s", PL[i].N);

        printf("Valor(R$): ");
        scanf("%f", &PL[i].Valor);

        Soma_PL += PL[i].Valor;
    }

    system("cls");

    printf("=========== RESULTADO BP ===========\n\n");

    printf("Total Ativos: %.2f\n", SA);
    printf("Total Passivos: %.2f\n", SP);
    printf("Total PL: %.2f\n", Soma_PL);

    printf("\nBP = %.2f\n\n", SA - (SP + Soma_PL));

    // =========================
    // DRE
    // =========================

    printf("=========== DRE ===========\n\n");

    printf("- Quantidade de receitas: ");
    scanf("%d", &QR);

    printf("- Quantidade de despesas: ");
    scanf("%d", &QD);

    Dados Receita[QR], Despesa[QD];

    system("cls");

    printf("=== RECEITAS ===\n");

    for(i = 0; i < QR; i++)
    {
        printf("Nome: ");
        scanf(" %[^\n]s", Receita[i].N);

        printf("Valor: ");
        scanf("%f", &Receita[i].Valor);
    }

    printf("\n=== DESPESAS ===\n");

    for(i = 0; i < QD; i++)
    {
        printf("Nome: ");
        scanf(" %[^\n]s", Despesa[i].N);

        printf("Valor: ");
        scanf("%f", &Despesa[i].Valor);
    }

    system("cls");

    Mostrar_DRE(QR, Receita, QD, Despesa);

    return 0;
}

// =========================
// FUNÇÕES
// =========================

void Mostrar_DRE(int R, Dados Receita[R], int D, Dados Despesa[D])
{
    int i;
    float SR = 0, SD = 0;

    printf("=======================\n");
    printf("          DRE          \n");
    printf("=======================\n\n");

    printf("Receitas:\n");
    for(i = 0; i < R; i++)
    {
        printf("- %s: %.2f\n", Receita[i].N, Receita[i].Valor);
        SR += Receita[i].Valor;
    }

    printf("\nTotal receitas: %.2f\n\n", SR);

    printf("Despesas:\n");
    for(i = 0; i < D; i++)
    {
        printf("- %s: %.2f\n", Despesa[i].N, Despesa[i].Valor);
        SD += Despesa[i].Valor;
    }

    printf("\nTotal despesas: %.2f\n\n", SD);

    float Resultado = SR - SD;

    printf("-----------------------\n");
    printf("Resultado: %.2f\n", Resultado);

    if(Resultado > 0)
        printf("Lucro\n");
    else if(Resultado < 0)
        printf("Prejuizo\n");
    else
        printf("Empate\n");

    printf("-----------------------\n");
}

int Maior_0(int V)
{
    if(V < 1)
        return 0;
    return 1;
}

void Titulo1()
{
    printf("=== ATIVOS ===\n\n");
}

void Titulo2()
{
    printf("=== PASSIVOS ===\n\n");
}

void Titulo3()
{
    printf("=== PATRIMONIO LIQUIDO ===\n\n");
}

void Circulante(int X, Dados B[X])
{
    int i;
    for(i = 0; i < X; i++)
        if(B[i].T == 1)
            printf("- %s: %.2f\n", B[i].N, B[i].Valor);
}

void Nao_Circulante(int X, Dados B[X])
{
    int i;
    for(i = 0; i < X; i++)
        if(B[i].T == 2)
            printf("- %s: %.2f\n", B[i].N, B[i].Valor);
}